teps/0121-refine-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,200 @@
+---
+status: proposed
+title: Refine Retries for TaskRuns and CustomRuns
+creation-date: '2022-09-08'
+last-updated: '2022-10-05'
+authors:
+- '@XinruZhang'
+---
+
+# TEP-0121: Refine Retries for TaskRuns and CustomRuns
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+- [Alternatives](#alternatives)
+- [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Two distinct imperfections on `Retries` we'd like to address in this TEP:
+- `Retries` on `Timeout` is designed inconsistently between TaskRun and CustomRun.
+  - For CustomRun, [the document](https://github.com/tektoncd/community/blob/33ca1d5254a405b1d479f2350443f6c7979a0b72/teps/0069-support-retries-for-custom-task-in-a-pipeline.md#proposal) instructs developers to **set `Timeout` for all retry attempts**. While the actual implementation is slightly different. See the [ref](https://github.com/tektoncd/pipeline/issues/5582).
+  - For TaskRun, the `Timeout` is **set for each retry attempt**. 
```

Comment:
Sorry @XinruZhang I am nitpicking because I am having hard time understanding some of the references and feel they are intermixed - I think we have four types - independent taskRun, taskRun as part of a pipeline, independent customRun, and customRun as part of a pipeline.

 This `taskRun` refers to `taskRun` created for a `pipelineTask`, right? `TaskRun` CRD specification (standalone TaskRun) does not support retries (which is being proposed as part of this TEP).

Today, a `timeout` specified as part of the `TaskRun` CRD specification applies to the one and only attempt. There are no multiple attempts supported for an independent TaskRun and hence this statement does not hold true in the context of independent taskRun CRD but applies to a taskRun when its part of a pipeline.

Reply:
Sorry for the confusion! Yes I'm referring to TaskRun created for a pipelineTask here. I updated some references to differentiate the concepts! Thanks!